Touch up the python docs (#418)
posit-dev/positron-docs: python-installations.qmd

---
title: "Discovering Python Installations"
description: "Understand how Positron discovers and manages Python installations. Covers automatic uv installation and support for uv, pyenv, conda, venv, and more environment managers."
---

This guide explains how Positron discovers and manages Python installations.

> For using Python in Positron, see our [guide to development](guide-python.qmd).

Reviewer: ```suggestion
For using Python in Positron, see our [Python language guide](guide-python.qmd).
```

Or something else maybe?

{{< include "_python-managers.qmd" >}}

Other tools might also be compatible, although Positron does not officially support them.

### Discovery locations

Positron searches for Python installations in the following locations.

> In general, it is best to use an environment that is highest in this list.

Reviewer 2: do we have an article to explain why this is

> it is best to use an environment that is highest in this list

maybe https://realpython.com/python-virtual-environments-a-primer/#why-do-you-need-virtual-environments ?

Author: I was actually thinking of making a blog post about this once my reorganization changes are in! Your link is a good resource in the meantime, so I can put it in.

* `.venv/` and `.conda/` folders in your [workspace](https://code.visualstudio.com/docs/editing/workspaces/workspaces#_singlefolder-workspaces) directory
* Tool-managed project-specific installations in locations like `~/.local/share/virtualenvs` or where `conda` stores envs

::: callout-important

> Virtual environments outside of the currently-opened folder will not be discovered.

Reviewer 2: This is true for project-level .venvs, but putting this alone in a callout sort of looks like no interpreters will be discovered outside of the current directory. 

I'd want to be really specific that "Positron will not discover project-level venvs that are declared in a project, unless you have that project open" (but in nicer wording)

## Automatic uv installation

If you do not have a valid Python available, Positron provides the option to install Python via [uv](https://docs.astral.sh/uv/) to provide an improved Python environment management experience.

> When Positron starts, it searches for Python on your machine and gives the option to install uv if you have no Python or only system Pythons available in the "Start New Console Session" dropdown.

Reviewer: ```suggestion
When Positron starts, it searches for Python on your machine and gives the option to install uv if you have no Python or only system Pythons available in the **Start New Console Session** dropdown.
```

I think we should bold this according to the style guide. Can you run the `/doc-reviewer` skill on this file to see how we're doing? Let me know if you've already done this and I am mistaken.

Author: I constantly forget to run this. 😂

The discovery process runs in the background and updates the interpreter list as new environments are found.

## Other details

### IPython kernel

Positron requires the [IPython kernel](https://pypi.org/project/ipykernel/) (version 6.19.1 or higher) to communicate with Python.
By default, the IPython kernel is bundled with Positron for the CPython implementation (the standard Python implementation) and added to the [Python search path](https://docs.python.org/3/library/sys_path_init.html).

> This means you don't have to manually install anything after creating a new environment, unlike working with notebooks in other tools.

Reviewer 2: ```suggestion
This means you don't have to manually install anything after creating a new environment.
```
a bit confusing to mention notebooks generally in a Python-forward doc

If you prefer finer control over your Python environment, set the [`python.useBundledIpykernel`](positron://settings/python.useBundledIpykernel) setting to false.
If you do that or are using an implementation other than CPython, install the IPython kernel yourself with `pip install ipykernel`.

### Unsupported Python versions

> Positron only supports Python minor versions 3.9 through 3.14, which are the [currently supported versions](https://devguide.python.org/versions/).

Reviewer 2: Hmmm...technically 3.9 is EOL, what do you think about saying something like "Positron supports the currently supported versions. It may continue to support newly EOL versions for a time after" OR just only mention supported versions and not guarantee/mention extra versions

- You open Positron
- You open a new workspace folder
- Positron detects certain file system changes in watched directories
- You refresh using the _Interpreter: Discover All Interpreters_ command

The discovery process runs in the background and updates the interpreter list as Positron finds new environments.

## Other details

### IPython kernel

> Positron requires the [IPython kernel](https://pypi.org/project/ipykernel/) (version 6.19.1 or higher) to communicate with Python.

Reviewer 2: i like this placement way more, thanks!
